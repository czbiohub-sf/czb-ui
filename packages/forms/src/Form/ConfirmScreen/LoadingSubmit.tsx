import React from "react";
import { HeadingSeparator } from "@czb-ui/core";
import { LoadingIndicator } from "@czi-sds/components";
import { Box } from "@mui/material";

export default function LoadingSubmit() {
  return (
    <div>
      <HeadingSeparator title="Submitting Sample..." />
      <Box my={4}>
        <LoadingIndicator sdsStyle="tag" />
      </Box>
    </div>
  );
}
