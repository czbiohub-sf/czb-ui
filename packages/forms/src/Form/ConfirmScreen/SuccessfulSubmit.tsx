import React from "react";
import { HeadingSeparator } from "@czb-ui/core";
import { Button } from "czifui";
import { Box, Typography } from "@mui/material";

interface SuccessfulSubmitProps {
  onSubmitAnotherResponse: () => void;
}

export default function SuccessfulSubmit({
  onSubmitAnotherResponse,
}: SuccessfulSubmitProps) {
  // TODO: Add prop to customize text?
  return (
    <div>
      <HeadingSeparator title="Sample submitted successfully!" />
      <Typography my={4}>Text here...</Typography>
      <Box my={4}>
        <Button
          sdsType="primary"
          sdsStyle="minimal"
          onClick={onSubmitAnotherResponse}
        >
          Submit another sample
        </Button>
      </Box>
    </div>
  );
}
