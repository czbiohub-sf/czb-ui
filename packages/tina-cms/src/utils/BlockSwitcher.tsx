import {
  GenericBannerBlock,
  GrandBannerBlock,
  TextBlock,
  InfoBoxBlock,
  GridBlock,
} from "../components";

interface BlockSwitcher {
  blocks: Array<any>;
}

export const BlockSwitcher = (props: BlockSwitcher) => {
  const blocks = props.blocks;

  return (
    <>
      {blocks
        ? blocks.map((block, i) => {
            switch (block.__typename) {
              case "PagesBlocksGenericBanner":
                return <GenericBannerBlock block={block} key={i} />;
              case "PagesBlocksGrandBanner":
                return <GrandBannerBlock block={block} key={i} />;
              case "PagesBlocksText":
                return <TextBlock block={block} key={i} />;
              case "PagesBlocksInfoBox":
                return <InfoBoxBlock block={block} key={i} />;
              case "PagesBlocksGrid":
                return <GridBlock block={block} key={i} />;
            }
          })
        : null}
    </>
  );
};
