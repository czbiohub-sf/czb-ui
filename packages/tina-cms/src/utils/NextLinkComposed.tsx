import * as React from "react";
import NextLink, { LinkProps as NextLinkProps } from "next/link";
import { styled } from "@mui/material/styles";

// Add support for the sx prop for consistency with the other branches.
const Anchor = styled("a")({});

interface NextLinkComposedProps
  extends Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, "href">,
    Omit<NextLinkProps, "href" | "as" | "onClick" | "onMouseEnter"> {
  to: NextLinkProps["href"];
  linkAs?: NextLinkProps["as"];
}

export const NextLinkComposed = React.forwardRef<
  HTMLAnchorElement,
  NextLinkComposedProps
>(function NextLinkComposed(props, ref) {
  const { to, linkAs, replace, scroll, shallow, prefetch, locale, ...other } =
    props;

  return (
    <NextLink
      href={to}
      prefetch={prefetch}
      as={linkAs}
      replace={replace}
      scroll={scroll}
      shallow={shallow}
      passHref
      locale={locale}
    >
      <Anchor ref={ref} {...other} />
    </NextLink>
  );
});
