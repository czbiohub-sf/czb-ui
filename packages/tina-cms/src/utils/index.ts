export * from "./NextLinkComposed";
export * from "./BlockSwitcher";
