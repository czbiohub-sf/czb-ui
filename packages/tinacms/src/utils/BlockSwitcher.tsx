import React from "react";
import {
  BannerBlock,
  GenericBannerBlock,
  GrandBannerBlock,
  TextBlock,
  InfoBoxBlock,
  TableBlock,
  LegacyInfoBoxBlock,
  HeadingSeparatorBlock,
  ImageGalleryBlock,
} from "../components";
import { Grid, Container } from "@mui/material";

// TODO: Fix types
interface BlockSwitcher {
  blocks: Array<any>;
  additionalBlocks?: any;
  disableContainerGutters?: boolean;
  disableYMargins?: boolean;
  smallVariants?: boolean;
}

export const BlockSwitcher = (props: BlockSwitcher) => {
  const blocks = props.blocks;

  const usableBlocks = {
    Banner: BannerBlock,
    GenericBanner: GenericBannerBlock,
    GrandBanner: GrandBannerBlock,
    Text: TextBlock,
    InfoBox: InfoBoxBlock,
    Grid: GridBlock,
    Table: TableBlock,
    LegacyInfoBox: LegacyInfoBoxBlock,
    HeadingSeparator: HeadingSeparatorBlock,
    ImageGallery: ImageGalleryBlock,
    ...props.additionalBlocks,
  };

  return (
    <>
      {blocks
        ? blocks.map((block, i) => {
            // Regex to remove all the collection name prefixes and the "Blocks" suffixes
            // Example block names: "PagesBlocksGenericBanner", "PagesBlocksGridBlocksInfoBox"
            // Outputs for those examples: "GenericBanner", "InfoBox"
            const blockToLookFor = block.__typename
              .replace(/(([A-Z][a-z0-9]+)+)Blocks/g, "")
              .replace(/Blocks/g, "");

            try {
              if (usableBlocks[blockToLookFor] === undefined) {
                throw new Error(
                  "Block not found. Is it supplied in the additionalBlocks prop?"
                );
              }
              return React.createElement(usableBlocks[blockToLookFor], {
                block,
                key: i,
                ...props,
              });
            } catch (e: any) {
              console.error(e);
              return (
                <div>
                  Something went wrong rendering the {blockToLookFor} block:{" "}
                  {e.message}
                </div>
              );
            }
          })
        : null}
    </>
  );
};

/*
 * Sample response
 * (what will be in the "block" prop of the GridBlock component)
 * {
 *	  "__typename": "PagesBlocksGrid",
 *	  "blocks": [
 *		  {
 *			  "__typename": "PagesBlocksGridBlocksInfoBox",
 *			  "title": "why tina cms is cool",
 *			  "subtitle": null,
 *			  ...
 *		  }
 *	  ]
 *	}
 */

interface GridBlockProps {
  __typename: string;
  blocks: Array<any>;
}

interface GridProps {
  block: GridBlockProps;
}

export const GridBlock = ({ block }: GridProps) => {
  const blocks = block.blocks;

  return (
    <Container sx={{ my: 8 }}>
      <Grid
        container
        spacing={7}
        columns={{ xs: 2, sm: 8, md: blocks?.length > 2 ? 12 : 8 }}
      >
        {blocks &&
          blocks.map((nestedBlock, i) => {
            return (
              <Grid item xs={2} sm={4} md={4} key={i}>
                <BlockSwitcher
                  blocks={[nestedBlock]}
                  disableContainerGutters
                  disableYMargins
                  smallVariants
                />
              </Grid>
            );
          })}
      </Grid>
    </Container>
  );
};
