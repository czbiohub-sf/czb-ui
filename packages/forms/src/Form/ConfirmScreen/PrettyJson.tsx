import { Typography, Box } from "@mui/material";
import { RJSFSchema } from "@rjsf/utils";
import React from "react";

interface PrettyJsonProps {
  formData: any[];
  schema: RJSFSchema[];
}

// Modification of https://stackoverflow.com/a/14810722/10013136
// Long story short, returns array instead of object in the original code
// k is key, v is value, and j is the index number (changed from i to not
// interfere with formData.map)
const objectMap = (
  obj: Record<string, any>,
  fn: (v: any, k: string, i: number) => any
) => Object.entries(obj).map(([k, v], j) => fn(v, k, j));

const isDataURL = (s: string) => {
  return /^data:text\/plain;name=\w+\.\w+;base64,[A-Za-z0-9+/]+=*$/.test(s);
};

const LayoutJson = (prop: string | string[] | Record<string, any>) => {
  if (typeof prop == "string" && isDataURL(prop)) {
    // It's a data url (from the file widget)
    return <div>test</div>;
  }

  if (typeof prop == "string") {
    return <div>{prop}</div>;
  }

  if (Array.isArray(prop)) {
    return (
      <ol>
        {prop.map((item, i) => (
          <li key={i}>{LayoutJson(item)}</li>
        ))}
      </ol>
    );
  }

  if (typeof prop == "object") {
    return (
      <ul>
        {objectMap(prop, (v, k, j) => {
          return (
            <Box mb={5} key={j}>
              <Typography fontWeight="bold">{k}</Typography>
              {LayoutJson(v)}
            </Box>
          );
        })}
      </ul>
    );
  }

  // If the type is not coded here,
  // just fallback to a JSON stringify
  return <div>{JSON.stringify(prop)}</div>;
};

// Actual schema is needed so we can get the real label names
// and also the form's page titles
export default function PrettyJson({ formData, schema }: PrettyJsonProps) {
  // For multi-step form
  return (
    <div>
      {formData.map((step: any, i) => {
        // "i" is the page/step of the form
        const schemaOnCurrentPage = schema[i];
        return (
          <div key={i}>
            <Typography variant="h2" my={4}>
              {schemaOnCurrentPage.title}
            </Typography>
            {objectMap(step, (v, k, j) => {
              const labelToShow = k || schemaOnCurrentPage.properties?.title;
              return (
                <Box mb={5} key={j}>
                  <Typography fontWeight="bold">{labelToShow}</Typography>
                  {LayoutJson(v)}
                </Box>
              );
            })}
          </div>
        );
      })}
    </div>
  );
}
