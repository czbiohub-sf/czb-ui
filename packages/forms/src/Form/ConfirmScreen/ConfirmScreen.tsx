import { Typography } from "@mui/material";
import { HeadingSeparator } from "@czb-ui/core";
import { Box } from "@mui/material";
import { RJSFSchema } from "@rjsf/utils";
import { Button } from "czifui";
import PrettyJson from "./PrettyJson/PrettyJson";

interface ConfirmScreenProps {
  formData: any;
  schema: RJSFSchema[];
  onConfirmation: () => void;
  onCancel: () => void;
  // TODO: onEditStep: (step: number) => void;
}

export default function ConfirmScreen({
  formData,
  schema,
  onConfirmation,
  onCancel,
}: ConfirmScreenProps) {
  return (
    <div>
      <HeadingSeparator title="Please confirm the data filled" />
      <PrettyJson formData={formData} schema={schema} />
      <Box my={8}>
        <Button
          sdsStyle="square"
          sdsType="secondary"
          sx={{ mr: 4 }}
          onClick={onCancel}
        >
          Cancel
        </Button>
        <Button
          sdsStyle="square"
          sdsType="primary"
          sx={{ mr: 4 }}
          onClick={onConfirmation}
        >
          Submit
        </Button>
      </Box>
    </div>
  );
}
