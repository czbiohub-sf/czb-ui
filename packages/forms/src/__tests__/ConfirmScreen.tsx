import renderer from "react-test-renderer";
import ConfirmScreen from "../Form/ConfirmScreen/ConfirmScreen";
import React from "react";
import { RJSFSchema } from "@rjsf/utils";

const completedFormProps = {
  formData: [
    {
      firstName: "Test",
      lastName: "Name",
      telephone: "1234567890",
    },
    {
      tasks: [
        {
          done: false,
          title: "Task 0",
          details: "I am task 0.",
        },
        {
          done: true,
          title: "Task 1",
          details: "I am task 1.",
        },
        {
          done: false,
          title: "Task 2",
          details: "I am task 2.",
        },
      ],
      title: "Task Title",
    },
    {
      files: [
        "data:text/plain;name=1Untitled.txt;base64,SSdtIGEgc21hbGwgdGV4dCBmaWxlLg==",
        "data:text/plain;name=2Untitled_copy%20-%20weird%20file%20name.txt;base64,SSdtIGFub3RoZXIgc21hbGwgdGV4dCBmaWxlLg==",
      ],
      file: "data:text/plain;name=1Untitled.txt;base64,SSdtIGEgc21hbGwgdGV4dCBmaWxlLg==",
    },
  ],
  schema: [
    {
      title: "A registration form",
      description:
        "Here is a form that almost has every input type we might need",
      type: "object",
      required: ["firstName", "lastName"],
      properties: {
        firstName: {
          type: "string",
          title: "First name",
          default: "Chuck",
        },
        lastName: {
          type: "string",
          title: "Last name",
        },
        telephone: {
          type: "string",
          title: "Telephone",
          minLength: 10,
        },
      },
    },
    {
      title: "A list of tasks",
      type: "object",
      required: ["title"],
      properties: {
        title: {
          type: "string",
          title: "Task list title",
        },
        tasks: {
          type: "array",
          title: "Tasks",
          items: {
            type: "object",
            required: ["title"],
            properties: {
              title: {
                type: "string",
                title: "Title",
                description: "A sample title",
              },
              details: {
                type: "string",
                title: "Task details",
                description: "Enter the task details",
              },
              done: {
                type: "boolean",
                title: "Done?",
                default: false,
              },
            },
          },
        },
      },
    },
    {
      title: "Test form",
      type: "object",
      properties: {
        file: {
          type: "string",
          format: "data-url",
          title: "Single file",
        },
        files: {
          type: "array",
          title: "Multiple files",
          items: {
            type: "string",
            format: "data-url",
          },
        },
        filesAccept: {
          type: "string",
          format: "data-url",
          title: "Single File with Accept attribute",
        },
      },
    },
  ] as RJSFSchema[],
};

it("renders form data correctly", () => {
  const component = renderer.create(
    <ConfirmScreen
      {...completedFormProps}
      onConfirmation={() => {
        1 + 1;
      }}
      onCancel={() => {
        1 + 1;
      }}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
