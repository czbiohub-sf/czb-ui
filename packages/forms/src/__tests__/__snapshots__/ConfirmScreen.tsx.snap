// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form data correctly 1`] = `
<div>
  <div
    className="MuiBox-root css-0"
  >
    <h1
      className="MuiTypography-root MuiTypography-h1 css-l4yn1p-MuiTypography-root"
    >
      Please confirm the data filled
    </h1>
    <div
      className="MuiBox-root css-168lvkp"
    />
  </div>
  <div>
    <div>
      <h2
        className="MuiTypography-root MuiTypography-h2 css-errr64-MuiTypography-root"
      >
        A registration form
      </h2>
      <div
        className="MuiBox-root css-10ib5jr"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          firstName
        </p>
        <div>
          Test
        </div>
      </div>
      <div
        className="MuiBox-root css-10ib5jr"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          lastName
        </p>
        <div>
          Name
        </div>
      </div>
      <div
        className="MuiBox-root css-10ib5jr"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          telephone
        </p>
        <div>
          1234567890
        </div>
      </div>
    </div>
    <div>
      <h2
        className="MuiTypography-root MuiTypography-h2 css-errr64-MuiTypography-root"
      >
        A list of tasks
      </h2>
      <div
        className="MuiBox-root css-10ib5jr"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          tasks
        </p>
        <ol>
          <li>
            <ul>
              <div
                className="MuiBox-root css-10ib5jr"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
                >
                  done
                </p>
                <div>
                  false
                </div>
              </div>
              <div
                className="MuiBox-root css-10ib5jr"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
                >
                  title
                </p>
                <div>
                  Task 0
                </div>
              </div>
              <div
                className="MuiBox-root css-10ib5jr"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
                >
                  details
                </p>
                <div>
                  I am task 0.
                </div>
              </div>
            </ul>
          </li>
          <li>
            <ul>
              <div
                className="MuiBox-root css-10ib5jr"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
                >
                  done
                </p>
                <div>
                  true
                </div>
              </div>
              <div
                className="MuiBox-root css-10ib5jr"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
                >
                  title
                </p>
                <div>
                  Task 1
                </div>
              </div>
              <div
                className="MuiBox-root css-10ib5jr"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
                >
                  details
                </p>
                <div>
                  I am task 1.
                </div>
              </div>
            </ul>
          </li>
          <li>
            <ul>
              <div
                className="MuiBox-root css-10ib5jr"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
                >
                  done
                </p>
                <div>
                  false
                </div>
              </div>
              <div
                className="MuiBox-root css-10ib5jr"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
                >
                  title
                </p>
                <div>
                  Task 2
                </div>
              </div>
              <div
                className="MuiBox-root css-10ib5jr"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
                >
                  details
                </p>
                <div>
                  I am task 2.
                </div>
              </div>
            </ul>
          </li>
        </ol>
      </div>
      <div
        className="MuiBox-root css-10ib5jr"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          title
        </p>
        <div>
          Task Title
        </div>
      </div>
    </div>
    <div>
      <h2
        className="MuiTypography-root MuiTypography-h2 css-errr64-MuiTypography-root"
      >
        Test form
      </h2>
      <div
        className="MuiBox-root css-10ib5jr"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          files
        </p>
        <ol>
          <li>
            <div>
              1Untitled.txt
            </div>
          </li>
          <li>
            <div>
              2Untitled_copy - weird file name.txt
            </div>
          </li>
          <li>
            <div>
              2Untitled_copy - i am a really long file nameeeeeeeeeeeeeeeeeee...
            </div>
          </li>
        </ol>
      </div>
      <div
        className="MuiBox-root css-10ib5jr"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          file
        </p>
        <div>
          1Untitled.txt
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiBox-root css-z727b5"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-hxp9em-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Cancel
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-18vxxcj-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Submit
    </button>
  </div>
</div>
`;
